crates/types/src/traits/block_contents.rs

```diff
@@ -50,7 +46,17 @@ pub trait BlockPayload:
     /// The type of the transitions we are applying
     type Transaction: Transaction;
 
-    // type Header: BlockHeader;
+    /// Data created during block building which feeds into the block header
+    type Metadata;
+
+    /// Build a payload and associated metadata with the transactions.
+    fn build(transactions: impl IntoIterator<Item = Self::Transaction>) -> (Self, Self::Metadata);
+
+    /// Encode the payload
+    fn encode(&self) -> Vec<u8>;
```

Comment:
Maybe lend out a reference instead of allocating new memory with each call.
```suggestion
    fn encode(&self) -> &[u8];
```

Reply:
I'm torn here. The latter forces the implementation to store the flat bytes as a field, so it doesn't even have the option of generating it on the fly. However, for the sequencer, we probably are going to do this anyways, so we have convenient access to the flat bytes for generating namespace proofs. In which case this interface would save a lot of copying.

There is a way around this, which allows the implementation to generate the flat bytes on the fly if desired, but doesn't require a copy if it already has them lying around. But it's not especially pretty:
```rust
fn encode<F, R>(&self, f: F) -> R
where
  F: FnOnce(&[u8]) ->R;

// Use like
let dispersal = block.encode(|bytes| vid_scheme.disperse(bytes));
```

Reply:
Ok I can't resist the call of the nerd snipe. You're going to need to explain that one to me. I don't see how your example relieves the implementation of the burden to store the flat bytes as a field. How is the implementation supposed to use `f`? Why does `f` take a `&[u8]`? Seems like the implementation still needs flat bytes to feed to `f`, no?

Reply:
An alternative is to return `impl Iterator<Iterm = u8>` and pray that compiler optimizations will eliminate any extra scans if the underlying type is something convenient like `&[u8]` or `Vec<u8>`.

Reply:
> I don't see how your example relieves the implementation of the burden to store the flat bytes as a field.

Ah, because the lifetime of the `&[u8]` in the argument of `F` is only required to live as long as the call to `f`. So the implementation can compute the flat byte array on the stack and pass that reference into `f`, instead of taking a reference from `&self`, like so:

```rust
fn encode<F, R>(&self, f: F) -> R {
  let flat_bytes = self.implementation_defined_flattening();
  f(&flat_bytes)
}
```

`impl Iterator<Item = u8>` would work nicely if `disperse` and `commit_only` took an iterator. But if they require a `&[u8]`, the caller is going to have to `collect()` that iterator into a `Vec<u8>` before it can be passed to VID, and I don't see how the compiler can eliminate that unless it's doing some really crazy inlining.

Reply:
> `impl Iterator<Item = u8>` would work nicely if `disperse` and `commit_only` took an iterator.

They once did. As I recall you recently convinced me not to bother taking an iterator in `disperse` and `commit_only`, so I changed it to `&[u8]` 😛 . But I could reinstate it without much difficulty. The only tiny difficulty is I need a convenient way to get the iterator length (so that payload reconstruction can truncate appropriately). I could use `ExactSizeIterator`, or I could hack the `bytes_to_field` function to return the number of bytes consumed.

Reply:
Yeah, I remember 🙃 I think I was thinking let's use the simplest type until we have a use case for something more general. And lo, a use case!

I think `ExactSizeIterator` might pose a problem, because the `encode` function may not know the exact size of the encoding without doing a full scan over the data, in which case why not just return a `Vec`? I don't really understand the suggestion about `bytes_to_field`.

Reply:
Ok this is a tangent unrelated to this PR (sorry Keyao).

The exact promise of `ExactSizeIterator` is not clear to me, and perhaps you're correct that it cannot promise not to do a linear scan of the data.

Moreover, `ExactSizeIterator` is not preserved by many common iterator compositions, which renders it a pretty useless trait. https://github.com/rust-lang/rust/pull/66531

To expound on `bytes_to_field`: This is an iterator adaptor that turns an iterator over `u8` to an iterator over `F`. It is possible to hack this function to return the number of bytes consumed, but any such hack would be an ergonomics nightmare because it can no longer be a pure iterator adaptor, so it's no longer composable with iterator design patterns.

Basically: I'm not aware of a clean solution here so I capitulated and enforced `&[u8]`.  The purist in me wants to make it an iterator but Rust needs a clean solution for this problem.

Reply:
@shenkeyao offline conversation with Jeb results in the following recommendation for `encode`:
```rust
trait BlockPayload {
  type Encode<'a>: 'a + Iterator<Item = u8>;

  fn encode(&self) -> Self::Encode<'_>;
}
```
Explanation:
- It's compatible with both methods of generating the flattened bytes: pre-stored or on-the-fly.
- Need the assoc type `Encode` because Rust has not yet stabilized impl trait return types in trait definitions

Reply:
@shenkeyao [more recent conversation with Jeb](https://github.com/EspressoSystems/hotshot-query-service/issues/266) suggests an even better solution: [`Cow`](https://doc.rust-lang.org/std/borrow/enum.Cow.html). Flagging this tip for future work.